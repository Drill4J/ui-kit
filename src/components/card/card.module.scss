@import '../../style-guide/index';

.card {
  width: 172px;
  height: 96px;
  padding: 16px;
  border: solid 1px map-get($monochrome, medium-tint);
  font-size: 12px;
  color: map-get($monochrome, default);

  &__methods-type {
    margin-left: 8px;
    font-family: $semibold-font;
  }

  &__total-methods {
    font-size: 20px;
    line-height: 32px;
    color: map-get($monochrome, black);
  }

  &__info-panel {
    line-height: 20px;

    :first-child {
      color: map-get($monochrome, black);
    }
  }

  &__risks {
    color: map-get($red, default);
  }

  &__coverage-bar {
    position: relative;
    display: flex;
    height: 8px;
    margin-bottom: 8px;
    background-color: map-get($monochrome, light-tint);
  }

  &__current-build {
    position: absolute;
    height: 8px;
    background-color: map-get($green, default);
  }

  &__active-scope {
    position: absolute;
    height: 8px;
    background-color: map-get($green, medium-tint);
  }

  &__overlapped-coverage {
    position: absolute;
    z-index: 2;
    height: 8px;
    background-color: map-get($green, shade);
  }

  &__coverage {
    font-family: $semibold-font;
    color: map-get($monochrome, black);
  }
}
