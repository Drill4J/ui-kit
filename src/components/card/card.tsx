import * as React from 'react';
import { BEM } from '@redneckz/react-bem-helper';

import { Panel } from '../../layouts';

import styles from './card.module.scss';

export interface Methods {
  totalCount?: number;
  coveredCount?: number;
  risksCount?: number;
  coveredFromTotalCount?: number;
  currentBuildCoverage?: number;
  overlappedCoverage?: number;
  activeScopeCoverage?: number;
  methodsType?: string;
}

interface Props {
  className?: string;
  icon: React.ReactNode;
  methods: Methods;
}

const card = BEM(styles);

export const Card = card(({
  className, icon, methods:
    {
      totalCount = 0, coveredCount = 0, risksCount = 0, coveredFromTotalCount = 0,
      currentBuildCoverage = 0,
      overlappedCoverage = 0,
      activeScopeCoverage = 0,
      methodsType = '',
    },
}: Props) => (
  <div className={className}>
    <Panel>
      {icon}
      <MethodsType>{methodsType}</MethodsType>
    </Panel>
    <TotalMethods>{totalCount}</TotalMethods>
    <InfoPanel align="space-between">
      <span>
        <Coverage>{coveredCount}</Coverage>
        {`/${totalCount}`}
      </span>
      {risksCount > 0 && (
        <Risks>{`${risksCount} risks`}</Risks>
      )}
    </InfoPanel>
    <CoverageBar>
      <CurrentBuild style={{ width: `${currentBuildCoverage}%` }} />
      <OverlappedCoverage style={{ width: `${overlappedCoverage}%`, marginLeft: `${currentBuildCoverage - overlappedCoverage}%` }} />
      <ActiveScope style={{ width: `${activeScopeCoverage}%`, marginLeft: `${currentBuildCoverage - overlappedCoverage}%` }} />
    </CoverageBar>
    <Panel align="space-between">
      <span>Covered</span>
      <Coverage>{`${coveredFromTotalCount}%`}</Coverage>
    </Panel>
  </div>
));

const MethodsType = card.methodsType('span');
const TotalMethods = card.totalMethods('div');
const InfoPanel = card.infoPanel(Panel);
const Risks = card.risks('span');
const CoverageBar = card.coverageBar('div');
const CurrentBuild = card.currentBuild('span');
const OverlappedCoverage = card.overlappedCoverage('span');
const ActiveScope = card.activeScope('span');
const Coverage = card.coverage('span');
