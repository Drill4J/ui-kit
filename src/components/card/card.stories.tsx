import React from 'react';
import { storiesOf } from '@storybook/react';

import { Card } from './card';
import { Icons } from '../icon';

storiesOf('Card', module)
  .add('Build coverage cards', () => (
    <div style={{ display: 'grid', gridTemplateColumns: '204px 204px 204px 204px', gridGap: '8px' }}>
      <Card
        icon={<Icons.Total />}
        methods={{
          totalCount: 2000,
          coveredCount: 600,
          risksCount: 640,
          coveredFromTotalCount: 30,
          currentBuildCoverage: 40,
          overlappedCoverage: 5,
          activeScopeCoverage: 10,
          methodsType: 'ALL METHODS',
        }}
      />
      <Card
        icon={<Icons.Add />}
        methods={{
          totalCount: 300,
          coveredCount: 40,
          risksCount: 260,
          coveredFromTotalCount: 2,
          currentBuildCoverage: 50,
          overlappedCoverage: 10,
          activeScopeCoverage: 10,
          methodsType: 'NEW',
        }}
      />
      <Card
        icon={<Icons.Edit height={15} width={16} viewBox="0 0 15 16" />}
        methods={{
          totalCount: 500,
          coveredCount: 120,
          risksCount: 380,
          coveredFromTotalCount: 6,
          currentBuildCoverage: 35,
          overlappedCoverage: 8,
          activeScopeCoverage: 40,
          methodsType: 'MODIFIED',
        }}
      />
      <Card
        icon={<Icons.Unaffected />}
        methods={{
          totalCount: 1200,
          coveredCount: 440,
          risksCount: 0,
          coveredFromTotalCount: 22,
          currentBuildCoverage: 40,
          overlappedCoverage: 15,
          activeScopeCoverage: 45,
          methodsType: 'UNAFFECTED',
        }}
      />
    </div>
  ));
