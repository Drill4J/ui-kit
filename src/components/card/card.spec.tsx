import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { Card } from './card';

describe('Card', () => {
  it('should match snapshot card', () => {
    const tree = renderer.create(<Card
      icon="Icon"
      methods={{
        totalCount: 2000,
        coveredCount: 600,
        risksCount: 640,
        coveredFromTotalCount: 30,
        currentBuildCoverage: 40,
        overlappedCoverage: 5,
        activeScopeCoverage: 10,
        methodsType: 'ALL METHODS',
      }}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display the card without risks, if there are none', () => {
    const wrapper = mount(<Card
      icon="Icon"
      methods={{
        totalCount: 500,
        coveredCount: 150,
        risksCount: 0,
        coveredFromTotalCount: 30,
        currentBuildCoverage: 40,
        overlappedCoverage: 5,
        activeScopeCoverage: 10,
        methodsType: 'UNAFFECTED',
      }}
    />);
    expect(wrapper.find('element(risks)')).toHaveLength(0);
  });
});
