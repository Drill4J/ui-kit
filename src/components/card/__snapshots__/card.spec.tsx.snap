// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot card 1`] = `
<div
  className="card"
>
  <div
    className="panel"
    onClick={[Function]}
  >
    Icon
    <span
      className="card__methods-type"
      data-modifiers=""
    >
      ALL METHODS
    </span>
  </div>
  <div
    className="card__total-methods"
    data-modifiers=""
  >
    2000
  </div>
  <div
    className="panel panel--align--space-between card__info-panel"
    onClick={[Function]}
  >
    <span>
      <span
        className="card__coverage"
        data-modifiers=""
      >
        600
      </span>
      /2000
    </span>
    <span
      className="card__risks"
      data-modifiers=""
    >
      640 risks
    </span>
  </div>
  <div
    className="card__coverage-bar"
    data-modifiers=""
  >
    <span
      className="card__current-build"
      data-modifiers=""
      style={
        Object {
          "width": "40%",
        }
      }
    />
    <span
      className="card__overlapped-coverage"
      data-modifiers=""
      style={
        Object {
          "marginLeft": "35%",
          "width": "5%",
        }
      }
    />
    <span
      className="card__active-scope"
      data-modifiers=""
      style={
        Object {
          "marginLeft": "35%",
          "width": "10%",
        }
      }
    />
  </div>
  <div
    className="panel panel--align--space-between"
    onClick={[Function]}
  >
    <span>
      Covered
    </span>
    <span
      className="card__coverage"
      data-modifiers=""
    >
      30%
    </span>
  </div>
</div>
`;
